# CADET v.0.2.1
Concatemer by Amplification DEtection Tool V.0.2.1


Whole Genome Amplification using multiple displacement amplification (MDA) sometimes can introduce potential false concatemer sequences that can affect whole genome assembly assays. Here we propose a Concatemer detection tool for those WGA assays.

<p align="center">
  
<img width="400" height="300" alt="CADET_AI" src="https://user-images.githubusercontent.com/28576450/206807841-2de5a0b3-4e00-460a-aaf1-34576894bf85.png">


</p>

**Figure. Concatemer-Mediated Multiple Displacement Amplification.** Annealing of random hexamer primers and addition of phi29-DNA polymerase leads to concatemers-mediated multiple displacement amplification from *(A)* linear and *(B)* circular concatemers respectively.  [Shoaib et al., 2008](https://bmcgenomics.biomedcentral.com/articles/10.1186/1471-2164-9-415).

## How it works?
It splits all reads in separate files to perform sliding windows with the user prefered size and the gap between these windows. For ONT amplified reads, we suggest windows >= 500bp with no overlaps (e.g. `-w 500` and `-s 500`). If the read is not able to generate more than one window (< 1kb in size in the 500bp window example) the read is skipped and its ID stored in the `short.txt` output file. Reads with more than two windows, will have their fragment windows aligned using nucmer and reads with overlaps are reported in stats and their IDs stored in the `concat_ID` output file. Statistics with the total number of reads, number of putative concatemers, number with no concatemer detection and the overlap frequency can be found in the `stats.txt` output.
Fastq/Fasta files with the characterized reads will also be generated.

  
### Workflow

  <p align="center">

<img width="500" height="500" alt="CADET_0 2 1" src="https://user-images.githubusercontent.com/28576450/206768044-3ed65ee6-b119-470c-bed7-a361d497efb4.png">

    
  </p>

## Instalation

Software Requirements:

- Mummer v3.23

- Seqtk v1.3

```
mamba create -n cadet -c bioconda -c conda-forge mummer seqtk 
```

## Usage
```
../CADET_v.0.2.1.sh [OPTIONS] -R <Reads.fastq> -w <window size> -s <slide size> -p <your_prefix>

Flag description:

Required:
  -R  --reads     fastq (or fasta*) file with reads generated by WGA sequencing using ONT (required)

Options: 
    -w  --window    length of desired window sequences in bp (required) (default = 500)
    -s  --slide     length to slide each window over in bp (required) (default = 500)
    -p  --prefix    Prefix name for your output folder (default = "CADET_output")  
    -h  --help      display this message

*Note: if using a FASTA file as input, the Concat.fastq output will be in fasta format
```

## Output Files
| Output File | Description |
| --- | --- |
|`concat_IDs`|sequence ID of putative concatemeric reads|
|`coords (folder)`| Folder with all nucmer coord alignments|
|`short.txt`| IDs of reads detected as short and skipped by the pipeline|
|`stats.txt`| File statistics of the CADET pipeline|
|`Non_conc.fastq`|fastq/fasta file containing non-concatemeric reads *(excludes short reads)*|
|`Conc.fastq`|fastq/fasta file containing putative concatemeric reads|
|`Short.fastq`|fastq/fasta file containing short reads|

### stats output example

Short sequences:                          1   

Number of non-concatemer detected:        3   **-------> Currently this number is considering non-concatemeric and short (will be fixed soon)**

Number of putative concatemer detected:   1   

Total number of Reads:                    4   

`###` putative concatemers `###`

read_number     read_ID self_alignemts

window_1  Long-sequence-one  3                -------> `read_number (useful for coords)` `read_name` `#_of_potential_concatemer_repeats`


